crates/sui-types/src/move_package.rs

```diff
@@ -154,9 +242,82 @@ impl MovePackage {
                 })
                 .collect(),
             max_move_package_size,
+            type_origin_table,
+            linkage_table,
         )
     }
 
+    fn build_linkage_table(
+        dependencies: &[MovePackage],
+    ) -> Result<BTreeMap<ObjectID, UpgradeInfo>, ExecutionError> {
+        // all transitive dependencies are included so simply put them into the table
+        let mut linkage_table = BTreeMap::new();
+        for p in dependencies {
+            let bytes: &Vec<u8> =
+                p.module_map.values().next().expect(
+                    "Tried to build a Move package from an empty iterator of Compiled modules",
+                );
+            let module = CompiledModule::deserialize(bytes)
+                .expect("A Move package contains a module that cannot be deserialized");
+            let original_id: ObjectID = (*module.address()).into();
+            linkage_table.insert(
+                original_id,
+                UpgradeInfo {
+                    upgraded_id: p.id,
+                    upgraded_version: p.version,
+                },
+            );
+        }
+        // TODO: verification
+        Ok(linkage_table)
+    }
+
+    fn build_initial_type_origin_table(
+        modules: &[CompiledModule],
+    ) -> BTreeMap<ModuleStruct, ObjectID> {
+        BTreeMap::from_iter(modules.iter().flat_map(|m| {
+            m.struct_defs().iter().map(|struct_def| {
+                let struct_handle = m.struct_handle_at(struct_def.struct_handle);
+                let module_name = m.name().to_string();
+                let struct_name = m.identifier_at(struct_handle.name).to_string();
+                let id: ObjectID = (*m.self_id().address()).into();
+                (
+                    ModuleStruct {
+                        module_name,
+                        struct_name,
+                    },
+                    id,
+                )
+            })
+        }))
+    }
+
+    fn build_upgraded_type_origin_table(
+        predecessor: &MovePackage,
+        modules: &[CompiledModule],
+    ) -> BTreeMap<ModuleStruct, ObjectID> {
+        let mut new_table = predecessor.type_origin_table.clone();
+        for m in modules {
+            for struct_def in m.struct_defs() {
+                let struct_handle = m.struct_handle_at(struct_def.struct_handle);
+                let module_name = m.name().to_string();
+                let struct_name = m.identifier_at(struct_handle.name).to_string();
+                let mod_struct = ModuleStruct {
+                    module_name,
+                    struct_name,
+                };
+                // only insert types that are not in the original table as only these should be
+                // marked as originating from the current package
+                if predecessor.type_origin_table.contains_key(&mod_struct) {
```

Comment:
Probably also worth adding an invariant violation if the predecessor contains a type that we don't have, because that means compatibility failed somehow.